### PR TITLE
Generalize InspectLedger and trace hard fork transitions

### PR DESCRIPTION
**Pull request by edsko**

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Fragment/Validated.hs [hunk 1]
```diff
@@ -10,11 +11,10 @@
 -- > import Ouroboros.Consensus.Fragment.Validated (ValidatedFragment)
 -- > import qualified Ouroboros.Consensus.Fragment.Validated as VF
 module Ouroboros.Consensus.Fragment.Validated (
-    ValidatedFragment -- Opaque
+    ValidatedFragment(ValidatedFragment)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Fragment/Validated.hs:14]: Why this change? So you can pattern match on it instead of using the accessors?

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Inspect.hs [hunk 2]
```diff
@@ -1,17 +1,59 @@
...
+partitionLedgerEvents ::
+     [LedgerEvent blk]
+  -> ([LedgerWarning blk], [LedgerUpdate blk])
+partitionLedgerEvents = partitionEithers . map ledgerEventToEither
+
+class ( Show     (LedgerWarning blk)
+      , Show     (LedgerUpdate  blk)
+      , Eq       (LedgerWarning blk)
+      , Eq       (LedgerUpdate  blk)
+      , Condense (LedgerUpdate  blk)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Inspect.hs:53]: Why `Condense` too?

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs [hunk 3]
```diff
@@ -519,71 +522,242 @@ data HardForkLedgerWarning xs =
     -- /before/ this lower bound, the node is misconfigured and will likely
     -- not work correctly. This should be taken care of as soon as possible
     -- (before the transition happens).
-  | HardForkLedgerWarningUnexpectedTransition EraParams EpochNo
+  | HardForkWarningTransitionMismatch (EraIndex xs) EraParams EpochNo
+
+    -- | Transition in the final era
+    --
+    -- The final era should never confirm any transitions
+  | HardForkWarningTransitionInFinalEra (EraIndex xs) EpochNo
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs:530]: This `(EraIndex xs)` is then the final era?

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs [hunk 4]
```diff
@@ -519,71 +522,242 @@ data HardForkLedgerWarning xs =
...
+
+    -- | An already-confirmed transition got un-confirmed
+  | HardForkWarningTransitionUnconfirmed (EraIndex xs)
+
+    -- | An already-confirmed transition got changed
+  | HardForkWarningTransitionReconfirmed (EraIndex xs) (EraIndex xs) EpochNo EpochNo
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs:536]: Document the arguments. I'm now guessing: transition from, to, was confirmed for epoch, is reconfirmed in this other epoch.

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs [hunk 5]
```diff
@@ -519,71 +522,242 @@ data HardForkLedgerWarning xs =
...
+                HardForkWarningTransitionUnconfirmed eraIndexZero
+            (Nil, Nothing, Just transition) ->
+              return $ LedgerWarning $
+                HardForkWarningTransitionInFinalEra eraIndexZero transition
+            (Nil, Just transition, Just transition') -> do
+              -- Only warn if the transition has changed
+              guard (transition /= transition')
+              return $ LedgerWarning $
+                HardForkWarningTransitionInFinalEra eraIndexZero transition
+            ((:*){}, Nothing, Just transition) ->
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs:625]: Inventing a new smiley along the way: `(:*){}`

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs [hunk 6]
```diff
@@ -519,71 +522,242 @@ data HardForkLedgerWarning xs =
...
+                         HardForkUpdateTransitionConfirmed
+                           eraIndexZero
+                           (eraIndexSucc eraIndexZero)
+                           transition
+                  else LedgerWarning $
+                         HardForkWarningTransitionMismatch
+                           eraIndexZero
+                           ps
+                           transition
+            ((:*){}, Just transition', Just transition) -> do
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs:638]: Is `transition'`, `transition` (note the `'`) intentional?

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/Serialisation/Golden.hs [hunk 7]
```diff
@@ -435,12 +434,8 @@ goldenTest_SerialiseNodeToClient codecConfig goldenDir Examples {..} =
...
 deriving instance ToExpr  TermToken
 
-instance Condense FlatTerm where
-  condense [] = "[]"
-  condense (x:ys) = intercalate "\n" $
-      [ "[ " <> show x ] <>
-      [ ", " <> show y | y <- ys ] <>
-      [ "]" ]
+instance Condense TermToken where
+  condense = show
```
Comment (mrBliss) [inline ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/Serialisation/Golden.hs:438]: Why this change? The previous instance printed each term on a separate line.

Update: I see you changed some things related to the `Condense` instance for lists (for condensing a list of events).

File: ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Address.hs [hunk 8]
```diff
@@ -1,17 +1,37 @@
...
 -- | Mock address
-type Addr = String
+newtype Addr = Addr String
+  deriving (
+      Show
+    , Eq
+    , Ord
+    , IsString
+    , Serialise
+    , NFData
```
Comment (mrBliss) [inline ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Address.hs:29]: Just wondering, where do we use the `NFData` instance?

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs [hunk 9]
```diff
@@ -335,9 +334,10 @@ protocolClientInfoShelley =
   Inspection
 -------------------------------------------------------------------------------}
 
+-- TODO: This should be updated as soon as we start preparing for the
+-- hard fork transition out of Shelley.
```
Comment (mrBliss) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs:338]: We should create a ticket for it with the right label.

File: ouroboros-consensus/src/Ouroboros/Consensus/Fragment/Validated.hs (hunk 1 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Fragment/Validated.hs:14]: Yup, exactly.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Inspect.hs (hunk 2 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Inspect.hs:53]: Used in the tests.

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs (hunk 3 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs:530]: Yes. Added "For clarity, we also record the index of that final era." to the docstring.

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs (hunk 4 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs:536]: Ok, wrong guess :) Added: "We record the indices of the era we are transitioning from and to, as well as the old and new 'EpochNo' of that transition, in that order.".

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs (hunk 6 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs:638]: No, well-spotted.

File: ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/Serialisation/Golden.hs (hunk 7 above)
Comment (edsko) [reply ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/Serialisation/Golden.hs:438]: Yeah, avoided overlapping instances for `Condense`. We needed to `condense` a `[LedgerUpdate blk]`, which is a type family, and hence could not resolve the overlapping instances.

File: ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Address.hs (hunk 8 above)
Comment (edsko) [reply ouroboros-consensus/ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Ledger/Address.hs:29]: I think we force elements in the mock ledger rather than enforcing it in the types.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs (hunk 9 above)
Comment (edsko) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs:338]: Created https://github.com/input-output-hk/ouroboros-network/issues/2452 , and informed Volodymyr that this should be part of the "prepare for next hard fork" item.

**Comment by edsko** [comment]
bors merge

**Comment by edsko** [comment]
bors merge

**Comment by edsko** [comment]
bors merge
